### ISSUE TITLE
saving nilable item with form

### ISSUE THREAD

**Issue by Djok39** [luckyframework/avram#27]
when save empty field it raises error

Should allow save nilable values in database
needed change in src/lucky_record/form.cr:
```
def set_{{ field[:name] }}_from_param(_value)
  parse_result = {{ field[:type] }}::Lucky.parse(_value)
  if parse_result.is_a? LuckyRecord::Type::SuccessfulCast
    {{ field[:name] }}.value = parse_result.value
  else
    {{ field[:name] }}.add_error "is invalid"
  end
end
```
to this:
```
def set_{{ field[:name] }}_from_param(_value)
  parse_result = {{ field[:type] }}::Lucky.parse(_value)
  if parse_result.is_a? LuckyRecord::Type::SuccessfulCast
    {{ field[:name] }}.value = parse_result.value
  else
    {% if field[:nilable] %}
      {{ field[:name] }}.value = nil
    {% else %}
      {{ field[:name] }}.add_error "is invalid"
    {% end %}
  end
end
```
or cast emty string as nil

**Comment by snadon** [comment]
I'm also having this issue and the suggest fix does work.

@paulcsmith Are you okay with this fix? I can make a PR.

**Comment by paulcsmith** [comment]
@snadon I think the problem with this approach is that if you want an Int32? and pass it `"123"` it would not leave an error, which it should. Instead I think the Lucky parser should treat `""` as `nil`. The logic for those is defined by each type and the base parser:

* https://github.com/luckyframework/avram/blob/master/src/avram/type.cr
* https://github.com/luckyframework/avram/tree/master/src/avram/charms

**Comment by snadon** [comment]
Okay, so rapidly, if we cast Nil whenever an Int32 as a value of "":
```
# src/avram/charms/int32_extensions.cr
# line: 10
def self.parse(value : String)
  if value.empty?
    SuccessfulCast(Nil).new(nil)
  else
    SuccessfulCast(Int32).new value.to_i
end
```
and to make it compile, we need to add `| Nil`:
```
# src/avram/where.cr
# line: 6
def initialize(@column : Symbol | String, @value : String | Array(String) | Array(Int32) | Nil)
end
```

This fixes my problem but I cannot run the tests at the moment. I would need to install docker on my laptop. Do you think there are repercussions?

If by any chance this is good, I can make a PR (including changes for Float64 and Int64).

**Comment by paulcsmith** [comment]
I think the change should be made in Avram::Type so that all object types get this behavior. Also adding the `Nil` to `Avram::Where` indicates maybe we need to handle it differently so that we don't leak `nil` to a bunch of different places. I'm not 100% sure how to approach it though since I haven't touched this part of the code.

If you wanna dive in more that is great, but if not I'll take a look later

**Comment by snadon** [comment]
Thank you @paulcsmith for your great help on this issue.

I'm dropping this here for anybody who needs a workaround:

```
# Add this file to your mixins folder
# src/forms/mixins/handle_nil_fields.cr
module HandleNilFields
  extend self

  def handle_nil(field)
    if field.value.nil? || field.param.empty?
      field.reset_errors
      field.value = nil
    end
  end
end
```

And use it like this:

```
# foo is any Int32? field that can be null in the database
# forms/test_form.cr
class TestForm < Test::BaseForm
  include HandleNilFields
  fillable foo

  def prepare
    handle_nil foo
  end
end
```

**Comment by jwoertink** [comment]
I believe this is fixed in master with the latest changes from #381.
